### PR TITLE
adding pickle mechanism without rootdir #182

### PR DESCRIPTION
**Pull request by dataisle**
carray.append works on objects without rootdir #182

``` python
import bcolz
b1 = bcolz.carray(["Hello", None])
b2 = bcolz.carray(["World"])
b1.append(b2)
b1
Out[4]:
carray((3,), object)
  nbytes: 199; cbytes: 631; ratio: 0.32
  cparams := cparams(clevel=5, shuffle=True, cname='blosclz')
[Hello None ['World']]
```
pickle works as well.
```python
import pickle
a=bcolz.carray([1,2,3])
e=pickle.dumps(a)
pickle.loads(e)
Out[7]:
carray((3,), int64)
  nbytes: 24; cbytes: 16.00 KB; ratio: 0.00
  cparams := cparams(clevel=5, shuffle=True, cname='blosclz')
[1 2 3]
```

### CODE REVIEW

File: bcolz/carray_ext.pyx [hunk 1]
```diff
@@ -2592,7 +2592,7 @@ cdef class carray:
 
     def __str__(self):
         return array2string(self)
-
+
```
Comment (esc) [inline bcolz/carray_ext.pyx]: Not sure about this, spurious whitespace inserted by accident?

File: bcolz/tests/test_carray.py [hunk 2]
```diff
@@ -77,6 +77,17 @@ def test_pickleable(self):
...
         self.assertEquals(b2.rootdir, b.rootdir)
 
+class pickleTestMemory(MayBeDiskTest, TestCase):
+    disk=False
+    def test_pickleable(self):
+        a = np.arange(1e2)
+        b = bcolz.carray(a, )
+        s = pickle.dumps(b)
+        if PY2:
+            self.assertTrue(type(s), str)
```
Comment (esc) [inline bcolz/tests/test_carray.py]: can you use ``assertIsInstance`` here?

File: bcolz/tests/test_carray.py [hunk 3]
```diff
@@ -77,6 +77,17 @@ def test_pickleable(self):
         b2 = pickle.loads(s)
         self.assertEquals(b2.rootdir, b.rootdir)
 
+class pickleTestMemory(MayBeDiskTest, TestCase):
+    disk=False
+    def test_pickleable(self):
+        a = np.arange(1e2)
```
Comment (esc) [inline bcolz/tests/test_carray.py]: You can use ```bcolz.arange``` here.

File: bcolz/tests/test_ctable.py [hunk 4]
```diff
@@ -1974,6 +1974,18 @@ def test_pickleable(self):
         self.assertEquals(b2.rootdir, b.rootdir)
         self.assertEquals(type(b2), type(b))
 
+    def test_pickleable_memory(self):
+        b = bcolz.ctable([[1, 2, 3], [1, 2, 3]],
+                         names=['a', 'b'],
+                         rootdir=None)
+        s = pickle.dumps(b)
+        if PY2:
+            self.assertTrue(type(s), str)
```
Comment (esc) [inline bcolz/tests/test_ctable.py]: Again ```assertIsInstance```.

**Comment by esc** [comment]
looks good in general, just a few points to address.

**Comment by esc** [comment]
also I'd like to list which issues this one resolves and I'll have to test it a little more.

**Comment by dataisle** [comment]
@esc Thanks! i added some of the idiomatic things that you mentioned in the pull request. As far as i can tell, this closes the bug reported in #182.

File: bcolz/tests/test_carray.py (hunk 3 above)
Comment (esc) [inline bcolz/tests/test_carray.py]: I meant, that you can use ```bcolz.arange(1e2)``` and ```bcolz.arange(1e2, rootdir=self.rootdir)``` directly.

Comment (esc) [inline bcolz/tests/test_carray.py]: Don't worry about it, I'll patch it up now.

Comment (dataisle) [inline bcolz/tests/test_carray.py]: thanks!

**Comment by esc** [comment]
Rebased in #194 closing this one.

**Pull request closed without merging**